hazelcast/src/main/java/com/hazelcast/sql/impl/expression/math/PlusFunction.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression.math;
+
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.SqlErrorCode;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.expression.BiExpressionWithType;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import com.hazelcast.sql.impl.type.QueryDataTypeFamily;
+
+import java.math.BigDecimal;
+
+/**
+ * Implements evaluation of SQL plus operator.
+ */
+public final class PlusFunction<T> extends BiExpressionWithType<T> implements IdentifiedDataSerializable {
+
+    public PlusFunction() {
+        // No-op.
+    }
+
+    private PlusFunction(Expression<?> operand1, Expression<?> operand2, QueryDataType resultType) {
+        super(operand1, operand2, resultType);
+    }
+
+    public static PlusFunction<?> create(Expression<?> operand1, Expression<?> operand2, QueryDataType resultType) {
+        return new PlusFunction<>(operand1, operand2, resultType);
+    }
+
+    @Override
+    public int getFactoryId() {
+        return SqlDataSerializerHook.F_ID;
+    }
+
+    @Override
+    public int getClassId() {
+        return SqlDataSerializerHook.EXPRESSION_PLUS;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public T eval(Row row, ExpressionEvalContext context) {
+        Object left = operand1.eval(row, context);
+        if (left == null) {
+            return null;
+        }
+
+        Object right = operand2.eval(row, context);
+        if (right == null) {
+            return null;
+        }
+
+        QueryDataTypeFamily family = resultType.getTypeFamily();
+        if (family.isTemporal()) {
+            throw new UnsupportedOperationException("temporal types are unsupported currently");
+        }
+
+        return (T) evalNumeric((Number) left, (Number) right, family);
+    }
+
+    private static Object evalNumeric(Number left, Number right, QueryDataTypeFamily family) {
+        switch (family) {
+            case TINYINT:
+                return (byte) (left.byteValue() + right.byteValue());
```

Comment:
A common comment to all math functions: how do we handle overflow? We are very careful with an overflow for `Converter`-s, but it seems that we may easily overflow here as well, and this is not handled properly. Or maybe the trick is that the overflow never happens here thanks type widening that happens during planning? If yes - is this the case for all math operators?

Can we add explicit tests to handle overflow for math operations to prove the correctness?

Reply:
Yes, the overflow should never happen if we didn't detect its possibility during validation and type inference. If its possibility is detected, the widening is performed until we hit `long`. Currently, all operations with `long` are checked for overflow, but that's not really necessary in all cases. It's just about equally expensive to check for an overflow as to check for the necessity of this check. But for compiled expressions that still might be a huge win to omit the overflow checks. See `HazelcastIntegerType.canOverflow`.